pecos/core/utils/mmap_util.hpp

```diff
@@ -0,0 +1,530 @@
+ * Must be incremented by 1 everytime when non-backward-compatible change is made. */
+#define __MMAP_UTIL_VER 1
+
+
+#include <cstring>
+#include <fcntl.h>
+#include <memory>
+#include <stdexcept>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <vector>
+#include <string>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include "file_util.hpp"
+
+
+namespace pecos {
+
+namespace mmap_util {
+
+namespace details_ { // namespace for Module Private classes
+
+/* Type constraint for mmap save/load.
+ * Please beware, this is only a minimum requirement and not a guarantee that the type is platform-portable for mmap.
+ * If the class/struct members need padding, please double-check mmap file portability on your own. */
+template<class Type>
+constexpr bool IsPlainOldData = std::is_trivially_copyable<Type>::value && std::is_standard_layout<Type>::value;
+
+template<typename Type, typename Ret=bool>
+using if_simple_serializable = std::enable_if_t<IsPlainOldData<Type>, Ret>;
+
+
+/*
+ * Module Private Class to store signature and metadata of all data blocks in mainbody for a memory-mapped file.
+ */
+class MmapMetadata_ {
+    public:
+        struct MetaInfo{
+            uint64_t offset; // Starting position of each block in the file
+            uint64_t size; // Number of bytes of each block's data
+            MetaInfo(uint64_t offset=0, uint64_t size=0) : offset(offset), size(size) {}
+        };
+
+        /* Save metadata to file.
+         * Parameters:
+         *      fp (FILE*) : File pointer.
+         *          Must points to the end of an opened file and waits for dumping metadata.
+         */
+        void save(FILE* fp) const {
+            uint64_t meta_offset = static_cast<uint64_t>(ftell(fp)); // Record metadata offset
+            Signature_ sig = Signature_(meta_offset);
+
+            file_util::fput_one<uint64_t>(info_.size(), fp);
+            file_util::fput_multiple<MetaInfo>(info_.data(), info_.size(), fp);
+
+            sig.save(fp);
+        }
+
+        /* Load metadata from file.
+         * Parameters:
+         *      fp (FILE*) : File pointer.
+         *          Load signature first, then rewind to load metadata.
+         */
+        void load(FILE* fp) {
+            Signature_ sig;
+            sig.load(fp);
+
+            fseek(fp, sig.meta_offset, SEEK_SET); // Rewind fp for loading metadata
+            uint64_t n_block = file_util::fget_one<uint64_t>(fp);
+            info_.resize(n_block);
+            file_util::fget_multiple<MetaInfo>(info_.data(), n_block, fp);
+        }
+
+        /* Get number of bytes to pad in order to align new block's beginning to multiple of N_ALIGN_BYTES_
+         * Parameters:
+         *      size (uint64_t): Number of bytes for the new block's data.
+         * Return:
+         *      n_bytes_to_pad (uint64_t): Number of bytes to pad before saving the new block.
+         */
+        uint64_t get_n_bytes_to_pad(const uint64_t size) const {
```

Comment:
There is not dependency on the input argument "size" in the entire implementation. Do we need this argument at all? if not, we can drop it and potentially rename the method to something like `get_num_of_padding_to_align(const uint64_t aligned_bytes=N_ALIGN_BYTES_) `